File: cicd_dagger_contract/python/validate.example.py
"""Validate module for Goserv"""

import json
from datetime import datetime
import dagger
from dagger import dag, function, object_type
from typing import Optional


@object_type
class Validate:
    @function
    async def validate(
        self,
        source: dagger.Directory,
        release_candidate: Optional[bool] = False,
        deployment_context: Optional[dagger.File] = None,
    ) -> dagger.File:
        """Validate runs the validation script to verify that the deployment is healthy and functioning correctly
        
        Args:
            source: Source directory containing the project
            release_candidate: Build as release candidate (appends -rc to version tag)
            deployment_context: Deployment context from Deploy function
        
        Returns:
            File containing validation context
        """
        # Extract deployment information from context if provided
        endpoint = None
        release_name = None
        if deployment_context:
            context_content = await deployment_context.contents()
            dep_context = json.loads(context_content)
            endpoint = dep_context.get("endpoint")
            release_name = dep_context.get("releaseName")

        # Perform validation checks
        # ... validation logic here ...

        # Create validation context
        validation_context = {
            "timestamp": datetime.now().isoformat(),
            "releaseName": release_name,
            "endpoint": endpoint,
            "status": "healthy",
            "healthChecks": ["pod-ready", "service-available"],
            "readinessChecks": ["http-200", "metrics-available"],
        }

        context_json = json.dumps(validation_context, indent=2)

        # Return as file
        return (
            dag.directory()
            .with_new_file("validation-context.json", context_json)
            .file("validation-context.json")
        )

File: cicd_dagger_contract/golang/validate.example.go
package main

import (
	"context"
	"encoding/json"
	"time"

	"dagger/goserv/internal/dagger"

	"dagger.io/dagger/dag"
)

// Validate runs the validation script to verify that the deployment is healthy and functioning correctly
func (m *Goserv) Validate(
	ctx context.Context,
	// Source directory containing the project
	source *dagger.Directory,
	// +optional
	// Build as release candidate (appends -rc to version tag)
	releaseCandidate bool,
	// +optional
	// Deployment context from Deploy function
	deploymentContext *dagger.File,
) (*dagger.File, error) {
	// Extract deployment information from context if provided
	var depContext map[string]interface{}
	if deploymentContext != nil {
		contextContent, err := deploymentContext.Contents(ctx)
		if err != nil {
			return nil, err
		}
		json.Unmarshal([]byte(contextContent), &depContext)
	}
	if err := json.Unmarshal([]byte(contextContent), &depContext); err != nil {
		return nil, err
	}

	endpoint := depContext["endpoint"].(string)
	releaseName := depContext["releaseName"].(string)

	// Perform validation checks
	// ... validation logic here ...

	// Create validation context
	validationContext := map[string]interface{}{
		"timestamp":       time.Now().Format(time.RFC3339),
		"releaseName":     releaseName,
		"endpoint":        endpoint,
		"status":          "healthy",
		"healthChecks":    []string{"pod-ready", "service-available"},
		"readinessChecks": []string{"http-200", "metrics-available"},
	}

	contextJSON, err := json.MarshalIndent(validationContext, "", "  ")
	if err != nil {
		return nil, err
	}

	// Return as file
	return dag.Directory().
		WithNewFile("validation-context.json", string(contextJSON)).
		File("validation-context.json"), nil
}
